src/Bicep.Core/Semantics/Namespaces/SystemNamespaceSymbol.cs

```diff
@@ -504,6 +505,46 @@ private static SyntaxBase StringLiteralFunctionReturnTypeEvaluator(FunctionCallS
             return SyntaxFactory.CreateStringLiteral(stringLiteral.RawStringValue);
         }
 
+        private static TypeSymbol JsonTypeBuilder(IBinder binder, IFileResolver fileResolver, IDiagnosticWriter diagnostics, ImmutableArray<FunctionArgumentSyntax> arguments, ImmutableArray<TypeSymbol> argumentTypes)
+        {
+            static TypeSymbol ToBicepType(JToken token)
+                => token switch {
+                    JObject @object => new ObjectType(
+                        "object",
+                        TypeSymbolValidationFlags.Default,
+                        @object.Properties().Select(x => new TypeProperty(x.Name, ToBicepType(x.Value), TypePropertyFlags.ReadOnly | TypePropertyFlags.ReadableAtDeployTime)),
+                        null),
+                    JArray @array => new TypedArrayType(
+                        UnionType.Create(@array.Select(x => ToBicepType(x))),
+                        TypeSymbolValidationFlags.Default),
+                    JValue value => value.Type switch {
+                        JTokenType.String => new StringLiteralType(value.ToString()),
+                        JTokenType.Integer => LanguageConstants.Int,
+                        JTokenType.Float => LanguageConstants.Int,
+                        JTokenType.Boolean => LanguageConstants.Bool,
+                        JTokenType.Null => LanguageConstants.Null,
+                        _ => LanguageConstants.Any,
+                    },
+                    _ => LanguageConstants.Any,
+                };
+
+            if (argumentTypes.Length != 1 || argumentTypes[0] is not StringLiteralType stringLiteral)
+            {
+                return LanguageConstants.Any;
+            }
+
+            try
+            {
+                var token = JToken.Parse(stringLiteral.RawStringValue);
+
+                return ToBicepType(token);
+            }
+            catch
+            {
+                return LanguageConstants.Any;
```

Comment:
There's probably a good argument for returning an error type with diagnostic attached here if we can be sure that the JSON is unparseable.

Reply:
`JToken.Parse()` takes settings:
![image](https://user-images.githubusercontent.com/22460039/124824824-bba7ff80-df27-11eb-9605-5afc6c434a67.png)

Should we explicitly create the settings object, so behavior never changes with Newtonsoft upgrades? 

Also, the default for duplicate properties is `Replace`, which is not what I expected. What does the runtime do if we pass in a JSON with duplicate properties to the `json()` function?

Reply:
Definitely agree about returning the error type, though.